pkg/cache/cache.go

```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"fmt"
+	"os"
+	"sync"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/tarball"
+	"github.com/spf13/afero"
+)
+
+const tarFmt string = "%s/%s.tar"
+
+// A Cache caches OCI images.
+type Cache interface {
+	Get(string, string) (v1.Image, error)
+	Store(string, string, v1.Image) error
+	Delete(string) error
+}
+
+// ImageCache stores and retrieves OCI images in a filesystem-backed cache in a
+// thread-safe manner.
+type ImageCache struct {
+	dir string
+	fs  afero.Fs
+	mu  sync.RWMutex
+}
+
+// NewImageCache creates a new ImageCache.
+func NewImageCache(dir string, fs afero.Fs) *ImageCache {
+	return &ImageCache{
+		dir: dir,
+		fs:  fs,
+	}
+}
+
+// Get retrieves an image from the ImageCache.
+func (c *ImageCache) Get(source, id string) (v1.Image, error) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	tag, err := name.NewTag(source)
+	if err != nil {
+		return nil, err
+	}
+	return tarball.ImageFromPath(fmt.Sprintf(tarFmt, c.dir, id), &tag)
+}
+
+// Store saves an image to the ImageCache.
+func (c *ImageCache) Store(source, id string, img v1.Image) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	ref, err := name.ParseReference(source)
+	if err != nil {
+		return err
+	}
+	cf, err := c.fs.Create(fmt.Sprintf(tarFmt, c.dir, id))
```

Comment:
What does this file path look like in practice? I have a bit of trouble following the naming logic; do we include the digest in the file path to ensure a cache lookup won't return an older version of a retagged image?

Reply:
The tarball is has the same name as the revision, so it would look something like `provider-aws-<truncated-image-digest>`. The digest ensures we won't get false cache hits. One thing to note here is that the `Package` controller would continue to look for the image at the given tag. This is not so different than the existing model, but I like to point it out because in the case that someone uses a tag like `latest`, if a reconcile gets triggered for the `Package` controller it will create a new `PackageRevision` and in the case where a user has `revisionActivationPolicy: Active`, the package manager will _automatically_ update to the new revision. If someone has `revisionActivationPolicy: Manual` the new revision will get created but we will not upgrade to it until they manually mark it as `Active`. So the safest policy would be to always use specific tags and set `revisionActivationPolicy: Manual`. However, in practice, just using specific tags will likely be sufficient as underlying contents should not be changing in a well-versioned project.